llvm/lib/Target/DirectX/DXILIntrinsicExpansion.cpp

```diff
@@ -197,6 +198,26 @@ static bool expandLog10Intrinsic(CallInst *Orig) {
   return expandLogIntrinsic(Orig, numbers::ln2f / numbers::ln10f);
 }
 
+static bool expandPowIntrinsic(CallInst *Orig) {
+
+  Value *X = Orig->getOperand(0);
+  Value *Y = Orig->getOperand(1);
+  Type *Ty = X->getType();
+  IRBuilder<> Builder(Orig->getParent());
+  Builder.SetInsertPoint(Orig);
+
+  auto *Log2Call =
```

Comment:
These edge cases may be addressed elsewhere, but do we want special cases for when X is 0?
Logs probably won't work when the input base parameter is 0, since it's mathematically undefined to take the log of 0. But, 0^3 should be 0 from a power op point of view. So I wonder if this implementation is sufficient for that edge case.

Reply:
Good question. So consider this implementation the base implementation without running any compiler optimizations. As such we can compare it to [DXC's TranslatePowImpl](https://github.com/microsoft/DirectXShaderCompiler/blob/main/lib/HLSL/HLOperationLower.cpp#L841C1-L846C78). So the first half of that is an fxc compat mode which I don't think we should carry forward. The second half of that is  exactly what we see in  https://godbolt.org/z/G5bKWz9j4 ie exp2(y*log2(x))

What you are refering to is something we could handle if
1.  we could determine X to be 0 at compile time
2.  if that[ constant could be propograted ](https://en.wikipedia.org/wiki/Constant_folding#Constant_propagation)
3. and then we could do CSE or a strength reduction  where we replace 0^3 -> 0

See example below DXC only returns 0 when both x and y in pow(x,y) are constants
https://godbolt.org/z/h595E7xP5

Reply:
This is an excellent, informative answer! Didn't know about constant propagation or strength reduction, very helpful resources :)
Presumably the optimization passes that would account for a 0 base would occur before this lowering occurs. If that's the case, LGTM

Reply:
@bob80905 I believe the current plan is to do optimizations after `dxil-op-lower`

Reply:
Ah, for a second I thought the `dxil-op-lower` pass would be final, it makes sense that optimizations could take place afterwards too. All good 👍